### PR TITLE
ASoC: SOF: Intel: remove mutual exclusion between HDA link and nocodec

### PR DESCRIPTION
**Pull request by plbossart**

### CODE REVIEW

File: sound/soc/sof/intel/hda.h [hunk 1]
```diff
@@ -414,7 +414,7 @@
 	(HDA_DSP_BDL_SIZE / sizeof(struct sof_intel_dsp_bdl))
 
 /* Number of DAIs */
-#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
```
Comment (ranj063) [inline sound/soc/sof/intel/hda.h:417]: commit message mismatch repeated

File: sound/soc/sof/intel/hda-bus.c [hunk 2]
```diff
@@ -70,9 +70,14 @@ void sof_hda_bus_init(struct snd_sof_dev *sdev, struct device *dev)
 {
 	struct hdac_bus *bus = sof_to_bus(sdev);
 
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA_LINK)
```
Comment (ranj063) [inline sound/soc/sof/intel/hda-bus.c:73]: grammatical error in the commit message "but for some tests, "we" need" or "but some tests"

File: sound/soc/sof/intel/Kconfig [hunk 3]
```diff
@@ -300,7 +300,6 @@ if SND_SOC_SOF_HDA_COMMON
 
 config SND_SOC_SOF_HDA_LINK
 	bool "SOF support for HDA Links(HDA/HDMI)"
-	depends on SND_SOC_SOF_NOCODEC_SUPPORT=n
```
Comment (ranj063) [inline sound/soc/sof/intel/Kconfig:303]: @plbossart does this mean we can also add the HDMI PCMs in the LNL nocodec topology?

**Review by ranj063 — COMMENTED** [review]
looks good @plbossart. just a couple minor nits regarding commit messages

**Review by RanderWang — COMMENTED** [review]
LGTM

File: sound/soc/sof/intel/Kconfig (hunk 3 above)
Comment (plbossart) [reply sound/soc/sof/intel/Kconfig:303]: no. What this means is that we can add SSP/DMIC with HDaudio link/DMA support in the nocodec build.

HDMI is still dependent on hdac_hda.c and that's controlled by the HDA_AUDIO_CODEC part.

HDAudio is a confusing thing, it means
a) a controller
b) a DMA
c) a link
d) a codec.

We support a) b) c) in nocodec mode.

File: sound/soc/sof/intel/hda.h (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda.h:417]: ah yes, fixed now

File: sound/soc/sof/intel/hda-bus.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-bus.c:73]: yes, fixed.

**Comment by plbossart** [comment]
v2: fixed commit messages and handled randconfig issue reported by kbuild bot

````diff
diff --git a/sound/soc/sof/intel/hda-dai.c b/sound/soc/sof/intel/hda-dai.c
index 5f4c0e87ec8e..8cb2dbc44f6c 100644
--- a/sound/soc/sof/intel/hda-dai.c
+++ b/sound/soc/sof/intel/hda-dai.c
@@ -241,14 +241,6 @@ static int hda_dai_hw_params(struct snd_pcm_substream *substream,
        return hda_dai_config(w, flags, &data);
 }
 
-static int hda_dai_prepare(struct snd_pcm_substream *substream, struct snd_soc_dai *dai)
-{
-       struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
-       int stream = substream->stream;
-
-       return hda_dai_hw_params(substream, &rtd->dpcm[stream].hw_params, dai);
-}
-
 /*
  * In contrast to IPC3, the dai trigger in IPC4 mixes pipeline state changes
  * (over IPC channel) and DMA state change (direct host register changes).
@@ -302,6 +294,16 @@ static int hda_dai_trigger(struct snd_pcm_substream *substream, int cmd, struct
        return 0;
 }
 
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA_AUDIO_CODEC)
+
+static int hda_dai_prepare(struct snd_pcm_substream *substream, struct snd_soc_dai *dai)
+{
+       struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+       int stream = substream->stream;
+
+       return hda_dai_hw_params(substream, &rtd->dpcm[stream].hw_params, dai);
+}
+
 static const struct snd_soc_dai_ops hda_dai_ops = {
        .hw_params = hda_dai_hw_params,
        .hw_free = hda_dai_hw_free,
@@ -309,6 +311,8 @@ static const struct snd_soc_dai_ops hda_dai_ops = {
        .prepare = hda_dai_prepare,
 };
 
+#endif
+

````

**Comment by plbossart** [comment]
back to draft since I am seeing issues with rmmod in local tests. something's not quite right still.

**Review by RanderWang — DISMISSED** [review]
LGTM

**Comment by plbossart** [comment]
Re-doing tests now that https://github.com/thesofproject/kconfig/pull/84 is merged

**Comment by plbossart** [comment]
SOFCI TEST

**Comment by plbossart** [comment]
@ujfalusi @ranj063 @bardliao can you please take a look?

**Review by bardliao — DISMISSED** [review]

**Comment by plbossart** [comment]
It looks like the CI build failed but the previous results are still shown. That's interesting. 

@fredoh9 @marc-hb @greg-intel @keqiaozhang how can I figure out what failed? There's something really wrong here.

**Comment by plbossart** [comment]
https://sof-ci.01.org/linuxpr/PR4289/build4280/build/index.html points to a non-existent blob and no replies -> redoing the checks.

**Comment by plbossart** [comment]
SOFCI TEST

**Comment by marc-hb** [comment]
> It looks like the CI build failed but the previous results are still shown. That's interesting.

That's unfortunate and should be changed but it's not a complete surprise if the first build was successful and the second was not.

Ideally the kernel SHA1s (and dates) should mitigate that issue (and are needed anyway)

File: sound/soc/sof/intel/hda.h (hunk 1 above)
Comment (ujfalusi) [reply sound/soc/sof/intel/hda.h:417]: I would have spelled out that the number of DAIs were based on CONFIG_SND_SOC_SOF_HDA while the declaration is using CONFIG_SND_SOC_SOF_HDA_AUDIO_CODEC, but that's just a nitpick.

File: sound/soc/sof/intel/hda-dai.c [hunk 4]
```diff
@@ -190,7 +190,8 @@ static int hda_link_dma_hw_params(struct snd_pcm_substream *substream,
 	return 0;
 }
 
-static int hda_dai_hw_free(struct snd_pcm_substream *substream, struct snd_soc_dai *cpu_dai)
+static int __maybe_unused hda_dai_hw_free(struct snd_pcm_substream *substream,
```
Comment (ujfalusi) [inline sound/soc/sof/intel/hda-dai.c:193]: type in commit message: 'defined **but** not used'

File: sound/soc/sof/intel/hda-dai.c [hunk 5]
```diff
@@ -208,9 +209,9 @@ static int hda_dai_hw_free(struct snd_pcm_substream *substream, struct snd_soc_d
 	return hda_link_dma_cleanup(substream, hext_stream, cpu_dai);
 }
 
-static int hda_dai_hw_params(struct snd_pcm_substream *substream,
-			     struct snd_pcm_hw_params *params,
-			     struct snd_soc_dai *dai)
+static int __maybe_unused hda_dai_hw_params(struct snd_pcm_substream *substream,
+					    struct snd_pcm_hw_params *params,
+					    struct snd_soc_dai *dai)
```
Comment (ujfalusi) [inline sound/soc/sof/intel/hda-dai.c:214]: Should not this be a scope of another series where the SSP/DMIC/SoundWire is added? But at that point the __maybe_unused can be dropped?

**Review by ujfalusi — COMMENTED** [review]
@plbossart, one typo and a question regarding to the __maybe_unused patch.

**Review by bardliao — DISMISSED** [review]

File: sound/soc/sof/intel/hda-dai.c (hunk 4 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-dai.c:193]: fixed, thanks

File: sound/soc/sof/intel/hda-dai.c (hunk 5 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-dai.c:214]: Whether this is done now or later, the point is that we first need to make some routines available and later use them. I don't want to make SSP/DMIC patches bigger just for a minor Kconfig 'hole' where we have unused stuff.

File: sound/soc/sof/intel/hda.h (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda.h:417]: I left the patch as is, it's just confusing to explain this Kconfig discrepancy.

**Review by ujfalusi — APPROVED** [review]

**Comment by plbossart** [comment]
@bardliao @RanderWang @ranj063 can I bother you to re-add your approval, I'd like to merge this tomorrow to rebase the DMIC/SSP stuff

**Review by RanderWang — APPROVED** [review]
LGTM

**Review by bardliao — APPROVED** [review]

**Comment by plbossart** [comment]
I can't see ANY of the test reports @greg-intel @marc-hb. All blobs are missing. what's going on?

https://sof-ci.01.org/linuxpr/PR4289/build4384/devicetest/index.html
https://sof-ci.01.org/linuxpr/PR4289/build4385/devicetest/index.html
https://sof-ci.01.org/linuxpr/PR4289/build4386/devicetest/index.html

**Comment by greg-intel** [comment]
@plbossart sorry for delay!  The reports are now available.  There was an outage due to the new VM lacking a specific ssh key. (We're migrating from a Shanghai physical machine.)

Issue: https://github.com/intel-innersource/drivers.audio.ci.sof-framework/issues/402

**Pull request merged**